subworkflows/nf-core/fastq_removeadapters_merge/main.nf

```diff
@@ -0,0 +1,148 @@
+        ch_reads                      = TRIMMOMATIC.out.trimmed_reads
+        ch_discarded_reads            = ch_discarded_reads.mix(TRIMMOMATIC.out.unpaired_reads.transpose()) // .transpose() because paired reads have 2 unpaired files in an array
+        ch_trimmomatic_trim_log       = TRIMMOMATIC.out.trim_log
+        ch_trimmomatic_summary        = TRIMMOMATIC.out.summary
+        ch_versions                   = ch_versions.mix(TRIMMOMATIC.out.versions.first())
+        ch_multiqc_files              = ch_multiqc_files.mix(TRIMMOMATIC.out.out_log)
+    }
+
+    if (!skip_cutadapt) {
+        CUTADAPT( ch_reads )
+        ch_reads         = CUTADAPT.out.reads
+        ch_multiqc_files = ch_multiqc_files.mix(CUTADAPT.out.log)
+    }
+
+    if (!skip_trimgalore) {
+        TRIMGALORE( ch_reads )
+        ch_reads               = TRIMGALORE.out.reads
+        ch_discarded_reads     = ch_discarded_reads.mix(TRIMGALORE.out.unpaired)
+        ch_trimgalore_log      = TRIMGALORE.out.log
+        ch_trimgalore_html     = TRIMGALORE.out.html
+        ch_trimgalore_zip      = TRIMGALORE.out.zip
+    }
+
+    if (!skip_bbduk) {
+        BBMAP_BBDUK( ch_reads, contaminants )
+        ch_reads         = BBMAP_BBDUK.out.reads
+        ch_versions      = ch_versions.mix(BBMAP_BBDUK.out.versions.first())
+        ch_multiqc_files = ch_multiqc_files.mix(BBMAP_BBDUK.out.log)
+    }
+
+    if (!skip_leehom) {
+        LEEHOM( ch_reads )
+
+        ch_reads = LEEHOM.out.fq_pass
+            .join(LEEHOM.out.unmerged_r1_fq_pass, by: 0, remainder: true)
+            .join(LEEHOM.out.unmerged_r2_fq_pass, by: 0, remainder: true)
+            .map { meta, single, r1, r2 ->
+                if (meta.single_end) {
+                    return [meta, single]
+                } else {
+                    return [meta, [r1, r2]]
+                }
+            }
+        ch_discarded_reads = ch_discarded_reads.mix(LEEHOM.out.fq_fail, LEEHOM.out.unmerged_r1_fq_fail, LEEHOM.out.unmerged_r2_fq_fail)
+        ch_versions        = ch_versions.mix(LEEHOM.out.versions.first())
+        ch_multiqc_files   = ch_multiqc_files.mix(LEEHOM.out.log)
+    }
+
+    if (!skip_fastp) {
+        FASTP(
+            ch_reads.map { meta, files ->  [ meta, files, contaminants ] },
+            fastp_discard_trimmed_pass,
+            fastp_save_trimmed_fail,
+            save_merged
+        )
+        ch_reads              = FASTP.out.reads
+        ch_merged_reads       = FASTP.out.reads_merged
+        ch_discarded_reads    = ch_discarded_reads.mix(FASTP.out.reads_fail.transpose()) // .transpose() because paired reads have 3 fail files in an array
+        ch_fastp_html         = FASTP.out.html
+        ch_fastp_log          = FASTP.out.log
+        ch_versions           = ch_versions.mix(FASTP.out.versions.first())
+        ch_multiqc_files      = ch_multiqc_files.mix(FASTP.out.json)
+    }
+
+    if (!skip_adapterremoval) {
+        ch_adapterremoval_in = ch_reads
```

Comment:
Another Question here just to double check, do all the other tools not care how they receive their data (i.e., can handle easily both)?

The reason why we have it separate for ADAPTEREMOVAL is more about the mkixing the output channels more than any thing else